sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveDirCommand.scala

```diff
@@ -90,6 +90,18 @@ case class InsertIntoHiveDirCommand(
         val dfs = qualifiedPath.getFileSystem(hadoopConf)
         (qualifiedPath, dfs)
       }
+
+    // SPARK-31459: If the target path already exists as a file,
+    // the last rename operation will cause only one result file to be moved,
+    // resulting in incorrect results.
+    if (fs.isFile(writeToPath)) {
+      if (overwrite) {
+        fs.delete(writeToPath, true)
```

Comment:
You might delete the target file too early. If the path to be written is a directory, its files/sub-directories are untouched when `saveAsHiveFile` fails. But if it is a file, you delete it before actually saving the query output.